### PR TITLE
fixed risky tests

### PR DESCRIPTION
**Pull request by fabpot**

### CODE REVIEW

File: test/Twig/Tests/NodeVisitor/OptimizerTest.php [hunk 1]
```diff
@@ -37,6 +37,8 @@ public function testRenderParentBlockOptimizer()
     public function testRenderVariableBlockOptimizer()
     {
         if (PHP_VERSION_ID >= 50400) {
+            $this->markTestSkipped('not needed on PHP >= 5.4');
+
             return;
```
Comment (stof) [inline test/Twig/Tests/NodeVisitor/OptimizerTest.php]: this return is now dead code

File: test/Twig/Tests/TemplateTest.php [hunk 2]
```diff
@@ -262,6 +266,8 @@ public function testGetAttributeStrict($defined, $value, $object, $item, $argume
             } catch (Twig_Error_Runtime $e) {
                 if (null !== $exceptionMessage) {
                     $this->assertSame($exceptionMessage, $e->getMessage());
+                } else {
+                    $this->addToAssertionCount(1);
```
Comment (stof) [inline test/Twig/Tests/TemplateTest.php]: much better would be to change this to use ``expectException`` and ``expectExpectionMessage`` (or ``setExpectedException('Twig_Error_Runtime', $exceptionMessage)`` on older versions) and remove the try/catch entirely

File: test/Twig/Tests/NodeVisitor/OptimizerTest.php (hunk 1 above)
Comment (fabpot) [inline test/Twig/Tests/NodeVisitor/OptimizerTest.php]: fixed

File: lib/Twig/Test/IntegrationTestCase.php [hunk 3]
```diff
@@ -121,6 +121,14 @@ public function getLegacyTests()
 
     protected function doIntegrationTest($file, $message, $condition, $templates, $exception, $outputs)
     {
+        if (!$outputs) {
+            // add a dummy assertion here to satisfy PHPUnit, it happens when
+            // we return a dummy test from getTests() when there is no legacy tests
+            $this->addToAssertionCount(1);
```
Comment (xabbuh) [inline lib/Twig/Test/IntegrationTestCase.php]: Doesn't `$this->markTestSkipped()` fit better here?

Comment (stof) [inline lib/Twig/Test/IntegrationTestCase.php]: yeah, probably

Comment (fabpot) [inline lib/Twig/Test/IntegrationTestCase.php]: much better indeed, fixed

File: test/Twig/Tests/TemplateTest.php (hunk 2 above)
Comment (fabpot) [inline test/Twig/Tests/TemplateTest.php]: Not that easy as we have some tests that throw exceptions without a message. Will see how to fix that later on.

Comment (fabpot) [inline test/Twig/Tests/TemplateTest.php]: fixed now

File: test/Twig/Tests/TemplateTest.php [hunk 4]
```diff
@@ -255,15 +259,17 @@ public function testGetAttributeStrict($defined, $value, $object, $item, $argume
...
             $this->assertEquals($value, $template->getAttribute($object, $item, $arguments, $type));
         } else {
-            try {
-                $this->assertEquals($value, $template->getAttribute($object, $item, $arguments, $type));
-
-                throw new Exception('Expected Twig_Error_Runtime exception.');
-            } catch (Twig_Error_Runtime $e) {
-                if (null !== $exceptionMessage) {
-                    $this->assertSame($exceptionMessage, $e->getMessage());
+            if (!$defined) {
```
Comment (stof) [inline test/Twig/Tests/TemplateTest.php]: this ``if`` will always be true, as it is inside the else clause of the inverse condition

Comment (fabpot) [inline test/Twig/Tests/TemplateTest.php]: indeed

**Pull request merged**